### PR TITLE
Parallelize job.peps_prs_report_df

### PR DESCRIPTION
**Pull request by danwinkler**
Before I jump too far into optimizing the inner loop of generating these pr curves, I realized that since we generate them for each run, we can trivially get a pretty good speedup by parallelizing over runs. In my largish test on my machine this made the `all_runs_pr_abund = job.peps_prs_report_df(**prs_args, pr_with_abundance=True)` line in the classify notebook go from taking around an hour (not sure exactly how long, never let it complete) to completing in around 5 mins.

I also added some comments with rough order of magnitude timing info about different parts of `pr_curve_by_pep_with_abundance`.

I also removed an assert in `ConfMat.from_true_pred`, which was taking upwards of 50ms in the inner loop. @zack-erisyon take a look at that and make sure it's okay.

I added `tqdm` for producing progress bars for `job.peps_prs_report_df`. `tqdm` supports progress bars in terminal and in notebooks, and detects the environment to know which one to produce. Example progress bar:
image

### CODE REVIEW

File: plaster/run/job.py [hunk 1]
```diff
@@ -59,12 +61,41 @@ def all_dfs(self, fn):
...
+            progress = tqdm(total=len(work_orders))
+
+            def progress_callback(i, j, retry):
+                if not retry:
+                    progress.update()
+
+            df_list = zap.work_orders(
+                work_orders,
+                _progress=progress_callback,
+                _cpu_limit=-4,  # TODO: what should this value be?
```
Comment (danwinkler) [inline plaster/run/job.py:88]: @zack-erisyon thoughts on how to set this value?

Comment (zack-erisyon) [reply plaster/run/job.py:88]: Yes -- do not set it.  Let it be controlled by the zap context.

File: plaster/run/job.py [hunk 2]
```diff
@@ -59,12 +61,41 @@ def all_dfs(self, fn):
...
         df_list = []
-        for run_i, run in enumerate(self._run_results.values()):
-            res_df = fn(run)
-            assert isinstance(res_df, pd.DataFrame)
-            res_df["run_i"] = run_i
-            res_df["run_name"] = run.manifest.run_name
-            df_list += [res_df]
+        if parallel:
+
+            def wrap_fn(run, run_i):
```
Comment (zack-erisyon) [inline plaster/run/job.py:66]: I'm not fuly awake, but did you look at zap's df_rows helper?

File: plaster/tools/utils/data.py [hunk 3]
```diff
@@ -117,7 +117,9 @@ def from_array(cls, arr):
     @classmethod
     def from_true_pred(cls, true, pred, true_dim, pred_dim):
         assert true.ndim == 1 and pred.ndim == 1
-        assert np.all((0 <= true) & (true < true_dim) & (0 <= pred) & (pred < pred_dim))
+        # This assert can take upwards of 50ms, and given that this fn is called as an inner loop, we can save some time if we omit it
+        # @ZBS, is it safe to remove this assert?
+        # assert np.all((0 <= true) & (true < true_dim) & (0 <= pred) & (pred < pred_dim))
```
Comment (zack-erisyon) [inline plaster/tools/utils/data.py:122]: Yes I think that is safe

**Review by zack-erisyon — APPROVED** [review]
Double check that the zap df_rows doesn't do what you want in that inner loop.
As for tqdm -- if we're going to use it then maybe we should in Pipeline too

File: plaster/run/job.py (hunk 2 above)
Comment (danwinkler) [reply plaster/run/job.py:66]: I might not be understanding what you're intending, but self._run_results isn't a dataframe, so I don't think df_rows is applicable.

**Pull request merged**